node/network/availability-distribution/src/lib.rs

```diff
@@ -0,0 +1,1947 @@
+) -> Result<HashMap<Hash, (Hash, CommittedCandidateReceipt)>>
+where
+	Context: SubsystemContext<Message = AvailabilityDistributionMessage>,
+{
+	let iter = relay_parents.into_iter();
+	let hint = iter.size_hint();
+
+	let capacity = hint.1.unwrap_or(hint.0) * (1 + AvailabilityDistributionSubsystem::K);
+	let mut live_candidates =
+		HashMap::<Hash, (Hash, CommittedCandidateReceipt)>::with_capacity(capacity);
+
+	for relay_parent in iter {
+
+		// register one of relay parents (not the ancestors)
+		let mut ancestors = query_up_to_k_ancestors_in_same_session(
+			ctx,
+			relay_parent,
+			AvailabilityDistributionSubsystem::K,
+		)
+		.await?;
+
+		ancestors.push(relay_parent);
+
+
+		// ancestors might overlap, so check the cache too
+		let unknown = ancestors
+			.into_iter()
+			.filter(|relay_parent_or_ancestor| {
+				// use the ones which we pulled before
+				// but keep the unknown relay parents
+				state
+					.receipts
+					.get(relay_parent_or_ancestor)
+					.and_then(|receipts| {
+						// directly extend the live_candidates with the cached value
+						live_candidates.extend(receipts.into_iter().map(
+							|(receipt_hash, receipt)| {
+								(
+									relay_parent,
+									(receipt_hash.clone(), receipt.clone()),
+								)
+							},
+						));
+						Some(())
+					})
+					.is_none()
+			})
+			.collect::<Vec<_>>();
+
+		// query the ones that were not present in the receipts cache
+		let receipts = query_live_candidates_without_ancestors(ctx, unknown.clone()).await?;
+		live_candidates.extend(
+			unknown.into_iter().zip(
+				receipts
+					.into_iter()
+					.map(|receipt| (receipt.hash(), receipt)),
+			),
+		);
+	}
+	Ok(live_candidates)
+}
+
+/// Query all para IDs.
+async fn query_para_ids<Context>(ctx: &mut Context, relay_parent: Hash) -> Result<Vec<ParaId>>
+where
+	Context: SubsystemContext<Message = AvailabilityDistributionMessage>,
+{
+	let (tx, rx) = oneshot::channel();
+	ctx.send_message(AllMessages::RuntimeApi(RuntimeApiMessage::Request(
+		relay_parent,
+		RuntimeApiRequest::AvailabilityCores(tx),
+	)))
+	.await
+	.map_err::<Error, _>(Into::into)?;
+
+	let all_para_ids: Vec<_> = rx
+		.await
+		.map_err::<Error, _>(Into::into)?
+		.map_err::<Error, _>(Into::into)?;
+
+	let occupied_para_ids = all_para_ids
+		.into_iter()
+		.filter_map(|core_state| {
+			if let CoreState::Occupied(occupied) = core_state {
+				Some(occupied.para_id)
+			} else {
+				None
+			}
+		})
+		.collect();
+	Ok(occupied_para_ids)
+}
+
+/// Modify the reputation of a peer based on its behavior.
+async fn modify_reputation<Context>(ctx: &mut Context, peer: PeerId, rep: Rep) -> Result<()>
+where
+	Context: SubsystemContext<Message = AvailabilityDistributionMessage>,
+{
+	trace!(
+		target: TARGET,
+		"Reputation change of {:?} for peer {:?}",
+		rep,
+		peer
+	);
+	ctx.send_message(AllMessages::NetworkBridge(
+		NetworkBridgeMessage::ReportPeer(peer, rep),
+	))
+	.await
+	.map_err::<Error, _>(Into::into)
```

Comment:
I think that i've managed to somehow dodge the explicit error conversions so far; but maybe there is a reason for this here

Reply:
It might have become obsolete, but at one point, it was required to make it compile, let me check if that is still mandatory.

Reply:
Some have, most didn't become obsolete.